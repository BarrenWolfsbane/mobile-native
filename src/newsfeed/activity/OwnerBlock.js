import React, {
  PureComponent
} from 'react';

import {
  NavigationActions
} from 'react-navigation';

import {
  Text,
  StyleSheet,
  TouchableOpacity,
  Image,
  View
} from 'react-native';

import {
  MINDS_URI
} from '../../config/Config';

/**
 * Owner Block Component
 */
export default class OwnerBlock extends PureComponent {

  state = {
    avatarSrc: { uri: MINDS_URI + 'icon/' + this.props.entity.ownerObj.guid + '/medium' }
  };

  /**
   * Navigate To channel
   */
  _navToChannel = () => {
    // only active if receive the navigation property
    if (this.props.navigation) {
      this.props.navigation.navigate('Channel', { guid:this.props.entity.ownerObj.guid});
    }
  }

  /**
   * Render
   */
  render() {
    const entity = this.props.entity.ownerObj;
    const rightToolbar = this.props.rightToolbar||null;

    return (
      <View style={styles.container}>
        <TouchableOpacity onPress={this._navToChannel}>
          <Image source={this.state.avatarSrc} style={styles.avatar}/>
        </TouchableOpacity>
        <View style={styles.body}>
          <TouchableOpacity onPress={this._navToChannel}>
            <Text style={styles.username}>
              { entity.username }
            </Text>
          </TouchableOpacity>
          {this.props.children}
        </View>
        {rightToolbar}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    padding: 8,
    alignItems: 'center',
    flexDirection: 'row',
  },
  avatar: {
    height: 46,
    width: 46,
    borderRadius: 23,
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: '#EEE',
  },
  body: {
    marginLeft: 8,
  },
  username: {
    fontWeight: 'bold',
  },
});