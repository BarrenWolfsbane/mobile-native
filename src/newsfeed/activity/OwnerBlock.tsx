import React, { PureComponent } from 'react';
import { Text, StyleSheet, TouchableOpacity, View } from 'react-native';
import FastImage from 'react-native-fast-image';

import withPreventDoubleTap from '../../common/components/PreventDoubleTap';
import ThemedStyles from '../../styles/ThemedStyles';
import type ActivityModel from '../ActivityModel';
import number from '../../common/helpers/number';
import i18nService from '../../common/services/i18n.service';
import { Icon } from 'react-native-elements';
const DebouncedTouchableOpacity = withPreventDoubleTap(TouchableOpacity);

type PropsType = {
  entity: ActivityModel;
  rightToolbar?: React.ReactNode;
  navigation: any;
  route?: any;
  children?: React.ReactNode;
};

/**
 * Owner Block Component
 */
export default class OwnerBlock extends PureComponent<PropsType> {
  /**
   * Navigate To channel
   */
  _navToChannel = () => {
    // only active if receive the navigation property
    if (this.props.navigation) {
      this.props.navigation.push('Channel', {
        guid: this.props.entity.ownerObj.guid,
        entity: this.props.entity.ownerObj,
      });
    }
  };

  /**
   * Navigate To group
   */
  _navToGroup = () => {
    if (this.props.navigation) {
      let groupGuid;
      try {
        groupGuid = this.props.route.params.group
          ? this.props.route.params.group.guid
          : this.props.route.params.guid;
      } catch {}
      if (
        this.props.entity.containerObj &&
        groupGuid === this.props.entity.containerObj.guid
      ) {
        return;
      }

      this.props.navigation.push('GroupView', {
        group: this.props.entity.containerObj,
      });
    }
  };

  get group() {
    if (!this.props.entity.containerObj) {
      return null;
    }

    return (
      <DebouncedTouchableOpacity
        onPress={this._navToGroup}
        style={styles.groupContainer}>
        <Text
          style={[styles.groupName, ThemedStyles.style.colorPrimaryText]}
          lineBreakMode="tail"
          numberOfLines={1}>
          > {this.props.entity.containerObj.name}
        </Text>
      </DebouncedTouchableOpacity>
    );
  }

  /**
   * Render
   */
  render() {
    const theme = ThemedStyles.style;
    const channel = this.props.entity.ownerObj;
    const rightToolbar = this.props.rightToolbar || null;

    const avatarSrc = channel.getAvatarSource();

    const showMetrics = !this.props.entity.boosted;

    return (
      <View style={styles.container}>
        <DebouncedTouchableOpacity onPress={this._navToChannel}>
          <FastImage source={avatarSrc} style={styles.avatar} />
        </DebouncedTouchableOpacity>
        <View style={styles.body}>
          <View style={styles.nameContainer}>
            <DebouncedTouchableOpacity onPress={this._navToChannel}>
              <Text
                style={[styles.username, ThemedStyles.style.colorPrimaryText]}>
                {channel.username}
              </Text>
            </DebouncedTouchableOpacity>
            {this.group}
          </View>
          {this.props.children}
        </View>
        {showMetrics ? (
          <Text
            style={[theme.marginRight2x, theme.colorTertiaryText, theme.fontM]}>
            {number(this.props.entity.impressions, 0)}{' '}
            {i18nService.t('views').toLowerCase()}
          </Text>
        ) : undefined}
        {this.props.entity.boosted ? (
          <View style={[theme.rowJustifyStart, theme.centered]}>
            <Icon
              type="ionicon"
              name="md-trending-up"
              size={18}
              style={theme.marginRight}
              color={ThemedStyles.getColor('tertiary_text')}
            />

            <Text
              style={[
                theme.marginRight2x,
                theme.colorTertiaryText,
                theme.fontM,
              ]}>
              {i18nService.t('boosted').toUpperCase()}
            </Text>
          </View>
        ) : undefined}
        {rightToolbar}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    padding: 10,
    alignItems: 'center',
    flexDirection: 'row',
  },
  avatar: {
    height: 52,
    width: 52,
    borderRadius: 26,
    borderWidth: 0.5,
    borderColor: '#EEE',
  },
  body: {
    marginLeft: 8,
    paddingRight: 36,
    flexWrap: 'wrap',
    flex: 1,
  },
  nameContainer: {
    flexDirection: 'row',
  },
  username: {
    fontWeight: 'bold',
    fontFamily: 'Roboto',
    color: '#444',
    fontSize: 16,
  },
  groupContainer: {
    marginLeft: 4,
    flex: 1,
  },
  groupName: {
    fontWeight: 'bold',
    fontFamily: 'Roboto',
  },
});
