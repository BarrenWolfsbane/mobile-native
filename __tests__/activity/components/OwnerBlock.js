import 'react-native';
import React from 'react';
import { Text, TouchableOpacity } from "react-native";
import { shallow } from 'enzyme';

import { activitiesServiceFaker } from '../../../__mocks__/fake/ActivitiesFaker';

import renderer from 'react-test-renderer';
import OwnerBlock from '../../../src/newsfeed/activity/OwnerBlock';
import withPreventDoubleTap from '../../../src/common/components/PreventDoubleTap';

// const DebouncedTouchableOpacity = withPreventDoubleTap(TouchableOpacity);
describe('Owner component', () => {
  
  let screen;
  beforeEach(() => {
    const TouchableOpacityCustom = <TouchableOpacity onPress={this.onPress} />;

    const navigation = { navigate: jest.fn(), push: jest.fn() };
    let activityResponse = activitiesServiceFaker().load(1);
    screen = shallow(
      <OwnerBlock entity ={activityResponse.activities[0]} navigation={navigation} rightToolbar={null}/>
    );

    jest.runAllTimers();
  });

  it('renders correctly', async () => {
    screen.update();
    expect(screen).toMatchSnapshot();
  });


  it('should have PreventDoubleTap', async () => {
    screen.update();
    expect(screen.find('PreventDoubleTap')).toHaveLength(2);
  });

  it('should _navToChannel on press ', () => {
    let activityResponse = activitiesServiceFaker().load(1);

    const navigation = {
      navigate: jest.fn(),
      push: jest.fn()
    };

    let entity = activityResponse.activities[0];
    entity.containerObj = { guid: 'guidguid' };
    screen = shallow(
      <OwnerBlock entity={entity} navigation={navigation} rightToolbar={null}/>
    );
    screen.update()
    let touchables = screen.find('PreventDoubleTap');
    touchables.at(0).props().onPress();
    jest.runAllTimers();

    expect(navigation.push).toHaveBeenCalledWith('Channel', {'entity': entity.ownerObj, 'guid': entity.ownerObj.guid});


    expect(screen.find('PreventDoubleTap')).toHaveLength(3);

  });


  it('should nav to groups on press ', () => {
    let activityResponse = activitiesServiceFaker().load(1);

    const navigation = {
      navigate: jest.fn(),
      push: jest.fn()
    };
    let entity = activityResponse.activities[0];
    entity.containerObj = { guid: 'guidguid' };
    screen = shallow(
      <OwnerBlock entity={entity} navigation={navigation} rightToolbar={null}/>
    );
    screen.update()
    let touchables = screen.find('PreventDoubleTap');

    expect(screen.find('PreventDoubleTap')).toHaveLength(3);
    //group touchable
    touchables.at(2).props().onPress();
    jest.runAllTimers();

    expect(navigation.push).toHaveBeenCalled();

  });


});
