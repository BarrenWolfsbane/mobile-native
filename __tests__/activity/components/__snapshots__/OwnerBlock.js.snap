// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Owner component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "padding": 8,
    }
  }
>
  <PreventDoubleTap
    onPress={[Function]}
  >
    <FastImage
      resizeMode="cover"
      source={
        Object {
          "headers": Object {
            "App-Version": "3.8.0",
            "Cache-Control": "no-cache, no-store, must-revalidate",
            "Pragma": "no-cache",
          },
          "uri": "https://cdn.minds.com/icon/824853017709780997/medium/undefined",
        }
      }
      style={
        Object {
          "borderColor": "#EEE",
          "borderRadius": 23,
          "borderWidth": 0.5,
          "height": 46,
          "width": 46,
        }
      }
    />
  </PreventDoubleTap>
  <View
    style={
      Object {
        "flex": 1,
        "flexWrap": "wrap",
        "marginLeft": 8,
        "paddingRight": 36,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <PreventDoubleTap
        onPress={[Function]}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#444",
                "fontFamily": "Roboto",
                "fontWeight": "bold",
              },
              undefined,
            ]
          }
        />
      </PreventDoubleTap>
      <PreventDoubleTap
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
            "marginLeft": 4,
          }
        }
      >
        <Text
          lineBreakMode="tail"
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontFamily": "Roboto",
                "fontWeight": "bold",
              },
              undefined,
            ]
          }
        >
          &gt; 
          group
        </Text>
      </PreventDoubleTap>
    </View>
  </View>
</View>
`;
